src/main/scala/mesosphere/marathon/core/storage/simple/SimplePersistenceStore.scala

```diff
@@ -0,0 +1,159 @@
+  private[this] val transactionMetric     = Metrics.counter(ServiceMetric, getClass, "transaction")
+  private[this] val transactionCountMetric = Metrics.counter(ServiceMetric, getClass, "transactionCount")
+
+
+  // Helper stages for logging and metrics
+  def logNode(prefix: String): Flow[Node, Node, NotUsed]                = Flow[Node].map{ n => logger.debug(s"$prefix${n.path}"); n }
+  def logPath(prefix: String): Flow[String, String, NotUsed]            = Flow[String].map{ p => logger.debug(s"$prefix${p}"); p }
+  def metric[T](counter: Counter, times: Long = 1): Flow[T, T, NotUsed] = Flow[T].map{ e => counter.increment(times); e}
+  // format: ON
+
+  override def create: Flow[Node, String, NotUsed] =
+    Flow[Node]
+      .via(logNode("Creating a node at "))
+      .via(metric(createMetric))
+      // `groupBy(path.hashCode % parallelism)` makes sure that updates to the same path always land in the same
+      // sub-stream, thus keeping the order of writes to the same path, even with parallelism > 1
+      .groupBy(parallelism, node => Math.abs(node.path.hashCode) % parallelism)
+      .map(node => factory
+        .create()
+        .forPath(node.path, node.data.toArray)
+        .toScala)
+      .mapAsync(parallelism)(identity)
+      .mergeSubstreams
+
+  override def read: Flow[String, Try[Node], NotUsed] =
+    Flow[String]
+      .via(logPath("Reading a node at "))
+      .via(metric(readMetric))
+      .map { path =>
+        factory
+          .getData()
+          .forPath(path)
+          .toScala
+          .map(bytes => Try(Node(path, ByteString(bytes))))
+          .recover {
+            case e: NoNodeException => Failure(e) // re-throw exception in all other cases
+          }
+      }
+      .mapAsync(parallelism)(identity)
+
+  override def update: Flow[Node, String, NotUsed] =
+    Flow[Node]
+      .via(logNode("Updating a node at "))
+      .via(metric(updatedMetric))
+      // `groupBy(path.hashCode % parallelism)` makes sure that updates to the same path always land in the same
+      // sub-stream, thus keeping the order of writes to the same path, even with parallelism > 1
+      .groupBy(parallelism, node => Math.abs(node.path.hashCode) % parallelism)
+      .map(node => factory
+        .setData()
+        .forPath(node.path, node.data.toArray)
+        .toScala
+        .map(_ => node.path))
+      .mapAsync(parallelism)(identity)
+      .mergeSubstreams
+
+  override def delete: Flow[String, String, NotUsed] =
+    Flow[String]
+      .via(logPath("Deleting a node at "))
+      .via(metric(deleteMetric))
+      .map(path => factory
+        .delete()
+        .forPath(path)
+        .toScala
+        .map(_ => path)) // Note: A Void is returned by the delete call and it can not be passed to the mapAsync stage
+      .mapAsync(parallelism)(identity)
+
+  override def children(path: String): Future[Seq[String]] = {
+    childrenMetric.increment()
```

Comment:
Why not `via(metric(...))`?

Reply:
Well, I don't create a `Source` or `Flow` here and `via` is a stream operator. The idea here is that in contrast to CRUD operations which offer stream interfaces because users might want to e.g. create a serious of nodes, `children`, `exists` and `sync` operations are typically used once so materializing a stream each time would be an unnecessary overhead.